File: leetcode/01249_minimum_remove_to_make_valid_parentheses.py
class Solution:
    def minRemoveToMakeValid(self, s: str) -> str:
        tmp = ''
        balance = 0
        for c in s:
            if c.isalpha() or c == '(':
                tmp += c
            balance += c == '('
            if c == ')' and balance:
                balance -= 1
                tmp += c
        answer = ''
        balance = 0
        for c in tmp[::-1]:
            if c.isalpha() or c == ')':
                answer += c
            balance += c == ')'
            if c == '(' and balance:
                balance -= 1
                answer += c
        return answer[::-1]

File: leetcode/01249_minimum_remove_to_make_valid_parentheses.rs
impl Solution {
    pub fn min_remove_to_make_valid(s: String) -> String {
        fn make_valid(z: String, open: char, close: char) -> String {
            let mut count = 0;
            let mut valid = String::new();
            for c in z.chars().rev() {
                count += (c == open) as i32 - (c == close) as i32;
                if (c == close && count < 0) { count += 1; continue; }
                valid.push(c);
            }
            valid
        }
        make_valid(make_valid(s, ')', '('), '(', ')')
    }
}

